test/gtest/core/test_allgather.cc

```diff
@@ -0,0 +1,162 @@
+                          ucc_dt_size(dtype) * count, mem_type));
+                ucc_mc_memcpy(coll->src.info.buffer, ctxs[r]->init_buf,
+                              ucc_dt_size(dtype) * count, mem_type,
+                              UCC_MEMORY_TYPE_HOST);
+            }
+        }
+        return ctxs;
+    }
+    void data_fini(UccCollCtxVec ctxs)
+    {
+        for (gtest_ucc_coll_ctx_t* ctx : ctxs) {
+            ucc_coll_args_t* coll = ctx->args;
+            if (coll->src.info.buffer) { /* no inplace */
+                UCC_CHECK(ucc_mc_free(coll->src.info.buffer, mem_type));
+            }
+            UCC_CHECK(ucc_mc_free(coll->dst.info.buffer, mem_type));
+            UCC_CHECK(ucc_mc_free(ctx->init_buf, UCC_MEMORY_TYPE_HOST));
+            free(coll);
+            free(ctx);
+        }
+        ctxs.clear();
+    }
+    void data_validate(UccCollCtxVec ctxs)
+    {
+        std::vector<uint8_t *> dsts(ctxs.size());
+
+        if (UCC_MEMORY_TYPE_HOST != mem_type) {
+            for (int r = 0; r < ctxs.size(); r++) {
+                UCC_CHECK(ucc_mc_alloc((void**)&dsts[r], ctxs[r]->rbuf_size,
+                                       UCC_MEMORY_TYPE_HOST));
+                ucc_mc_memcpy(dsts[r], ctxs[r]->args->dst.info.buffer,
+                              ctxs[r]->rbuf_size, UCC_MEMORY_TYPE_HOST, mem_type);
+            }
+        } else {
+            for (int r = 0; r < ctxs.size(); r++) {
+                dsts[r] = (uint8_t *)(ctxs[r]->args->dst.info.buffer);
+            }
+        }
+        for (int i = 0; i < ctxs.size(); i++) {
+            uint8_t *rbuf = dsts[i];
+            for (int r = 0; r < ctxs.size(); r++) {
+                size_t rank_size = ucc_dt_size((ctxs[r])->args->src.info.datatype) *
+                        (ctxs[r])->args->src.info.count;
+                for (int i = 0; i < rank_size; i++) {
+                    EXPECT_EQ(r, rbuf[r*rank_size + i]);
+                }
+            }
+        }
+        if (UCC_MEMORY_TYPE_HOST != mem_type) {
+            for (int r = 0; r < ctxs.size(); r++) {
+                ucc_mc_free((void*)dsts[r], UCC_MEMORY_TYPE_HOST);
+            }
+        }
+    }
+};
+
+class test_allgather_0 : public test_allgather,
+        public ::testing::WithParamInterface<Param_0> {};
+
+UCC_TEST_P(test_allgather_0, single_host)
+{
+    const int size = std::get<0>(GetParam());
+    const ucc_datatype_t dtype = (ucc_datatype_t)std::get<1>(GetParam());
+    const int count = std::get<2>(GetParam());
+    const gtest_ucc_inplace_t inplace = std::get<3>(GetParam());
+
+    UccTeam_h team = UccJob::getStaticJob()->create_team(size);
+    set_inplace(inplace);
+    set_mem_type(UCC_MEMORY_TYPE_HOST);
+    UccCollCtxVec args = data_init(size, dtype, count);
+    UccReq    req(team, args);
+    req.start();
+    req.wait();
+    data_validate(args);
+    data_fini(args);
+}
+
+#ifdef HAVE_CUDA
+UCC_TEST_P(test_allgather_0, single_cuda)
```

Comment:
Looks exactly the same as previous test. I think it's better to add memory type as parameter and pass it to INSTANTIATE_TEST_CASE_P

Reply:
fixed